aws/resource_aws_gamelift_game_session_queue.go

```diff
@@ -0,0 +1,204 @@
+package aws
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/gamelift"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"log"
+)
+
+func resourceAwsGameliftGameSessionQueue() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsGameliftGameSessionQueueCreate,
+		Read:   resourceAwsGameliftGameSessionQueueRead,
+		Update: resourceAwsGameliftGameSessionQueueUpdate,
+		Delete: resourceAwsGameliftGameSessionQueueDelete,
+
+		Schema: map[string]*schema.Schema{
+			"destinations": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 128),
+			},
+			"player_latency_policy": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"maximum_individual_player_latency_milliseconds": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntAtLeast(0),
+						},
+						"policy_duration_seconds": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(0),
+						},
+					},
+				},
+			},
+			"timeout_in_seconds": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(10, 600),
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsGameliftGameSessionQueueCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+
+	input := gamelift.CreateGameSessionQueueInput{
+		Name:                  aws.String(d.Get("name").(string)),
+		Destinations:          expandGameliftGameSessionQueueDestinations(d.Get("destinations").([]interface{})),
+		PlayerLatencyPolicies: expandGameliftGameSessionPlayerLatencyPolicies(d.Get("player_latency_policy").([]interface{})),
+		TimeoutInSeconds:      aws.Int64(int64(d.Get("timeout_in_seconds").(int))),
+	}
+	log.Printf("[INFO] Creating Gamelift Session Queue: %s", input)
+	out, err := conn.CreateGameSessionQueue(&input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.GameSessionQueue.GameSessionQueueArn)
```

Comment:
Since the `Read`, `Update`, and `Delete` functions can all work off the `name` attribute, I'd recommend switch setting the ID to that and referencing via `d.Id()` elsewhere. If/when we want to support resource import, then we can simply use this in the schema definition:

```go
Importer: &schema.ResourceImporter{
	State: schema.ImportStatePassthrough,
},
```

Tested by adding a `TestStep` like this to each acceptance test:

```go
{
  ResourceName:      resourceName,
  ImportState:       true,
  ImportStateVerify: true,
},
```

And document via:

``````markdown
## Import

Gamelift Game Session Queues can be imported by their `name`, e.g.

```
$ terraform import aws_gamelift_game_session_queue.example example
```
``````

Reply:
Is it ok to use `name` as ID even if `name` can be changed?

Reply:
Can the `name` be changed? 😄 Usually they have a `NewName` or something parameter in the update API call (or `Name` with some other more specific ID), but I'm not seeing that here: https://docs.aws.amazon.com/gamelift/latest/apireference/API_UpdateGameSessionQueue.html

If you're correct though, then we would likely want to stick with ARN. 👍

Reply:
Yeah, it’s just the `Name`. It can be changed both using UI or API and it is even changing in the test case.

Reply:
Though what about the import in this case? I don’t mind to implement it as well. If I understand correctly we should implement import by ARN?

Reply:
The attribute in Terraform appears to be correctly marked `ForceNew: true` so it should be deleting the old queue and recreating the new one. When I create a queue in the UI, the option to edit the queue name is grayed out. Am I missing something?

![screen shot 2018-11-14 at 12 42 47 am](https://user-images.githubusercontent.com/189114/48462527-85ca1180-e7a6-11e8-91a5-8cbb5bf60feb.png)

Reply:
Oh, sorry! It’s my bad. I will replace ARN by name soon. Thank you!